scripts/src/public/services/orders.ts

```diff
@@ -63,103 +69,114 @@ function orderDbToApi(order: db.Order, logger: LoggerInstance): Order {
+		expiration_date: order.expirationDate!.toISOString(),
 	};
 }
 
 export async function submitOrder(
 	orderId: string, form: string | undefined, walletAddress: string, appId: string, logger: LoggerInstance): Promise<Order> {
 
-	const openOrder = await db.OpenOrder.findOneById(orderId);
-	if (!openOrder) {
-		throw Error(`no such order ${orderId}`);
+	const order = await db.Order.findOneById(orderId);
+	if (!order) {
+		throw Error(`no such open order ${ orderId }`);
 	}
-	if (new Date() > openOrder.expiration) {
-		throw Error(`order ${orderId} expired`);
+	if (new Date() > order.expirationDate!) {
+		throw Error(`open order ${ orderId } has expired`);
 	}
 
-	const offer = await offerDb.Offer.findOneById(openOrder.offerId);
+	const offer = await offerDb.Offer.findOneById(order.offerId);
 	if (!offer) {
-		throw new Error(`no such offer ${ openOrder.offerId }`);
+		throw Error(`no such offer ${ order.offerId }`);
 	}
-
 	if (offer.type === "earn") {
 		// validate form
 		if (!await offerContents.isValid(offer.id, form, logger)) {
-			throw Error(`submitted form is invalid for ${openOrder.id}`);
+			throw Error(`submitted form is invalid for ${ order.id }`);
 		}
 	}
 
-	// transition open order to pending order
-	const order = db.Order.new({
-		id: openOrder.id,
-		userId: openOrder.userId,
-		offerId: openOrder.offerId,
-		amount: offer.amount,
-		type: offer.type,
-		status: "pending",
-		meta: offer.meta.order_meta,
-	});
-	offer.cap.used += 1;
-	await offer.save();
+	order.status = "pending";
+	order.currentStatusDate = new Date();
 	await order.save();
-	await openOrder.delete();
 
-	// pay or start timer for payment
 	if (offer.type === "earn") {
 		await payment.payTo(walletAddress, appId, offer.amount, order.id, logger);
-	} else {
-		await submitSpend(order, offer, walletAddress, appId, logger);
 	}
-	metrics.submitOrder(offer.type, offer.id);
 
-	return orderDbToApi(order, logger);
-}
-
-export async function submitSpend(
-	order: db.Order, offer: offerDb.Offer, walletAddress: string, appId: string, logger: LoggerInstance): Promise<void> {
-	async function makeFailed(orderId: string) {
-		// XXX lock on order.id
-		const order = await db.Order.findOneById(orderId);
-		if (!order) {
-			throw new Error(`no order ${ orderId }`);
-		}
-
-		order.status = "failed";
-		const offer = await offerDb.Offer.findOneById(order.offerId);
-		if (!offer) {
-			throw new Error(`no offer ${ order.offerId }`);
-		}
-
-		offer.cap.used -= 1;
-		await offer.save();
-		await order.save();
-	}
-
-	// start a timer for order.expiration + grace till this order becomes failed
-	// setTimeout(makeFailed, order.expiration, order.id);
-	return;
+	metrics.submitOrder(offer.type, offer.id);
+	return orderDbToApi(order);
 }
 
 export async function cancelOrder(orderId: string, logger: LoggerInstance): Promise<void> {
 	// you can only delete an open order - not a pending order
-	// validate order
+	const order = await db.Order.findOneById(orderId);
```

Comment:
you can only cancel opened:
`db.Order.findOne({id: orderId, status: "opened"});`